nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/ExtractEmailHeaders.java

```diff
@@ -168,21 +173,40 @@ public void process(final InputStream rawIn) throws IOException {
+                        recipients = InternetAddress.parseHeader(originalMessage.getHeader(Message.RecipientType.TO.toString(), ","), false);
+                        for (int toCount = 0; toCount < ArrayUtils.getLength(recipients); toCount++) {
+                            attributes.put(EMAIL_HEADER_TO + "." + toCount, recipients[toCount].toString());
                         }
-                        for (int toCount = 0; toCount < ArrayUtils.getLength(originalMessage.getRecipients(Message.RecipientType.BCC)); toCount++) {
-                            attributes.put(EMAIL_HEADER_BCC + "." + toCount, originalMessage.getRecipients(Message.RecipientType.BCC)[toCount].toString());
+                    }
+                    if (originalMessage.getHeader(Message.RecipientType.BCC.toString(), ",") != null) {
+                        recipients = InternetAddress.parseHeader(originalMessage.getHeader(Message.RecipientType.BCC.toString(), ","), false);
+                        for (int toCount = 0; toCount < ArrayUtils.getLength(recipients); toCount++) {
+                            attributes.put(EMAIL_HEADER_BCC + "." + toCount, recipients[toCount].toString());
                         }
-                        for (int toCount = 0; toCount < ArrayUtils.getLength(originalMessage.getRecipients(Message.RecipientType.CC)); toCount++) {
-                            attributes.put(EMAIL_HEADER_CC + "." + toCount, originalMessage.getRecipients(Message.RecipientType.CC)[toCount].toString());
+                    }
+                    if (originalMessage.getHeader(Message.RecipientType.CC.toString(), ",") != null) {
+                        recipients = InternetAddress.parseHeader(originalMessage.getHeader(Message.RecipientType.CC.toString(), ","), false);
+                        for (int toCount = 0; toCount < ArrayUtils.getLength(recipients); toCount++) {
+                            attributes.put(EMAIL_HEADER_CC + "." + toCount, recipients[toCount].toString());
                         }
                     }
-                    // Incredibly enough RFC-2822 specified From as a "mailbox-list" so an array I returned by getFrom
-                    for (int toCount = 0; toCount < ArrayUtils.getLength(originalMessage.getFrom()); toCount++) {
-                        attributes.put(EMAIL_HEADER_FROM + "." + toCount, originalMessage.getFrom()[toCount].toString());
+
+                    // Get Non-Strict Sender Addresses
+                    InternetAddress[] sender = null;
+                    if (originalMessage.getHeader("From",",") != null) {
+                        sender = (InternetAddress[])ArrayUtils.addAll(sender, InternetAddress.parseHeader(originalMessage.getHeader("From", ","), false));
+                    }
+                    if (originalMessage.getHeader("Sender",",") != null) {
+                        sender = (InternetAddress[])ArrayUtils.addAll(sender, InternetAddress.parseHeader(originalMessage.getHeader("Sender", ","), false));
```

Comment:
I don't think we should combine the values of the Sender and From headers here, for two reasons: 

1. It breaks the documentation for the processor that states email.headers.from will contain the values in From
2. If From is a mailbox-list, Sender might be a single mailbox in the From mailbox-list, so the result here could be a duplicate email address.

I think this could break expectations from folks already using the ExtractEmailHeaders processor. If this functionality is needed, a new header field (ie, email.headers.sender) should be added. However, in that case, this functionality could already be achieved using the Additional Headers capability offered by ExtractEmailHeaders.

Reply:
My logic here was that I wanted ALL of the From/Sender addresses. So mailbox-list or not, in sender or not, this would collect them all. Note that I'm merging them. So if they are both present, then they will both be added.

Again, I'll have to re-read the RFC to see if this is correct. Based on the implementation of getFrom() I found on grepcode though, I figured it was. 

Perhaps a bad assumption though, because having read [RFC 822](https://www.ietf.org/rfc/rfc822.txt) a lot of implementations get email addresses wrong. I've seen plenty of accepted mail like " "@example.com I think the break-down is in the SHOULD/MUST contract, where in a mail server SHOULD accept that address.

Let me read up on the additional RFCs and get back to you. I can also do some digging in my mail archive to see what postfix has accepted/interpreted in the past. I've seen a lot of email address regexes that break because they don't assume "this is a valid address"@example.com is valid, even though postfix accepted it.

Reply:
I do have an example of a from address that was previously rejected and postfix accepted. It looked like this:

From: Automated System <<automated-system@example.com>>

Reply:
Thanks @btwood. I'll look into the getFrom() implementation. I didn't think of that, but you're right that the best thing to do would try to match that implementation as closely as possible as that is what was being called before. 

I thought about non-strict parsing more this morning. Would it be possible to make strict vs. non-strict parsing a user configurable toggle in the processor configuration? If so, and if the default was strict = true, then that would address my concern about breaking existing data flows for NiFi users if/when they upgrade.

Reply:
It absolutely would be possible. Instead of setting it manually to "false", we just need a variable set by the processor config. And pass that to getHeader(header, strict). I'm not certain how to accomplish that yet, but I'll look into it.

My day job is catching up with me today. I don't have free cycles at work at the moment that allow me to work on this. I will be free later this afternoon though.

Reply:
Cool, no rush. I've got other things to work on today as well. Will get back to this sometime over the weekend or next week.

Reply:
Also, I agree. We don't want to break anything for current users. If the expected behavior changes, we don't know how that impacts current users, because some may be relying on this to fail.